src/devices/nm-device.c

```diff
@@ -15977,17 +15977,21 @@ nm_device_spec_match_list_full (NMDevice *self, const GSList *specs, int no_matc
 {
 	NMDeviceClass *klass;
 	NMMatchSpecMatchType m;
+	const char *hw_address = NULL;
 
 	g_return_val_if_fail (NM_IS_DEVICE (self), FALSE);
 
 	klass = NM_DEVICE_GET_CLASS (self);
 
+	if (!nm_device_get_unmanaged_flags (self, NM_UNMANAGED_PLATFORM_INIT))
+		hw_address = nm_device_get_permanent_hw_address (self);
```

Comment:
the problem is, that we read the permanent MAC address via ethtool, which does look ups by name, which is racing with udev renaming the device. That means, we must not try to use ethtool before NM_UNMANAGED_PLATFORM_INIT.

But that poses a problem if we want to make decisions for a device-spec *before* NM_UNMANAGED_PLATFORM_INIT. 

For software devices this is no real problem, because they anyway don't have a real permenent MAC adddress. I guess, you should also consider the `out_is_fake` argument of `nm_device_get_permanent_hw_address_full()` and maybe ignore it. Or maybe not. Dunno... WDYT?

Reply:
> instead, call `nm_device_get_permanent_hw_address_full()` where you can select whether to freeze.

Do you mean something like:
```
   hw_address = nm_device_get_permanent_hw_address_full (self,
                                                         !nm_device_get_unmanaged_flags (self, NM_UNMANAGED_PLATFORM_INIT),
                                                         &is_fake);
   m = nm_match_spec_device (specs,
                             ...
                             is_fake ? NULL : hw_address,

```
? Yeah, it should be equivalent.

Reply:
exactly like this. Except I am undecided what to do about `@is_fake`. Should such addresses be ignored (like in this snippet) or not. Probably they should be ignored. WDYT?

Reply:
Yes, I think the matching function should consider only the real MAC address.